pmd-visualforce/src/main/java/net/sourceforge/pmd/lang/vf/rule/security/AbstractVfTypedElExpressionRule.java

```diff
@@ -0,0 +1,89 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.vf.rule.security;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+
+import net.sourceforge.pmd.RuleContext;
+import net.sourceforge.pmd.lang.vf.IdentifierType;
+import net.sourceforge.pmd.lang.vf.VfExpressionTypeVisitor;
+import net.sourceforge.pmd.lang.vf.ast.ASTCompilationUnit;
+import net.sourceforge.pmd.lang.vf.ast.ASTIdentifier;
+import net.sourceforge.pmd.lang.vf.rule.AbstractVfRule;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+import net.sourceforge.pmd.properties.PropertyFactory;
+
+/**
+ * Represents a rule where the {@link net.sourceforge.pmd.lang.vf.ast.ASTIdentifier} nodes are enhanced with the
+ * node's {@link IdentifierType}. This is achieved by processing metadata files referenced by the Visualforce page.
+ */
+abstract class AbstractVfTypedElExpressionRule extends AbstractVfRule {
+    /**
+     * Directory that contains Apex classes that may be referenced from a Visualforce page.
+     */
+    private static final PropertyDescriptor<List<String>> APEX_DIRECTORIES_DESCRIPTOR =
```

Comment:
@oowekyala https://github.com/pmd/pmd/issues/2518 looks promising and solves the issue I originally ran into when trying to store the info on the AST and needed to add configuration properties.

2518 is currently targeted for 7.0. What are your thoughts on an interim 6.x solution which
1. Adds the CLI extension `-L<langId>:<propName> <value>` 
2. Stores the CLI values either in `ParserOptions` or in a globally available singleton

This doesn't solve the lifecycle issues, but it does provide a means for language specific properties in a generic way.

Thanks.

Reply:
> What are your thoughts on an interim 6.x solution which
>  1. Adds the CLI extension -L<langId>:<propName> <value>
>  2. Stores the CLI values either in ParserOptions or in a globally available singleton

I think this could work! I don't expect this part of the proposal to change (the CLI syntax), and anyway it would be experimental. Implementing it early will also make it easier for this PR to be adapted to language properties when they're implemented. I think it would be easier to touch ParserOptions than making a global singleton. For one, this would give a way to set those properties in tests. It's also what #2830 does.

In this PR, as an interim to the interim, I suppose you could make the changes you need to ParserOptions, and configure it via system properties. This way you can pass values to the properties without needing to implement the new CLI syntax here. We can make a PR next, to replace those system properties with the new CLI syntax.

Reply:
@oowekyala I updated the PR to have ParserOptions extend AbstractPropertySource. Let me know your thoughts. If this is too big of a change for 6.x, I can change VFParserOptions to implement PropertySource and delegate to a concrete instance of AbstractPropertySource.